Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java

```diff
@@ -0,0 +1,306 @@
+                externalContext.responseSendError(HttpServletResponse.SC_NOT_FOUND,
+                        ((HttpServletRequest) externalContext.getRequest()).getRequestURI());
+            } else {
+                externalContext.responseSendError(HttpServletResponse.SC_NOT_FOUND, null);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private void process(StreamedContent streamedContent, ExternalContext externalContext) throws IOException {
+        externalContext.setResponseStatus(HttpServletResponse.SC_OK);
+
+        if (Objects.nonNull(streamedContent.getContentEncoding())) {
+            externalContext.setResponseHeader("Content-Encoding", streamedContent.getContentEncoding());
+        }
+
+        // Adapt implementation of Warren Dew
+        // (https://stackoverflow.com/questions/28427339/how-to-implement-http-byte-range-requests-in-spring-mvc)
+        // using org.primefaces.application.resource.StreamedContentHandler
+        HttpServletResponse response = (HttpServletResponse) externalContext.getResponse();
+
+        if (Objects.nonNull(streamedContent.getName())) {
+            response.setHeader("Content-Disposition", "inline;filename=\"" + streamedContent.getName() + "\"");
+            response.setHeader("ETag", streamedContent.getName());
+        }
+
+        response.setHeader("Accept-Ranges", "bytes");
+        response.setBufferSize(DEFAULT_BUFFER_SIZE);
+
+        processInputStreamToOutputStream((HttpServletRequest) externalContext.getRequest(), response, streamedContent,
+                externalContext.getResponseOutputStream());
+    }
+
+    private void processInputStreamToOutputStream(HttpServletRequest request, HttpServletResponse response,
+            StreamedContent streamedContent, OutputStream outputStream) throws IOException {
+        InputStream inputStream = streamedContent.getStream();
+        // Prepare some variables. The full Range represents the complete file.
+        int length = inputStream.available(); // Length of file
+        Range full = new Range(0, length - 1, length);
+        List<Range> ranges = getRanges(request, response, length, streamedContent.getName());
+
+        if (ranges.isEmpty() || Objects.equals(ranges.get(0), full)) {
+            // Return full file.
+            LOGGER.info("Return full file");
+            response.setContentType(streamedContent.getContentType());
+            response.setHeader("Content-Range", "bytes " + full.start + "-" + full.end + "/" + full.total);
+            response.setHeader("Content-Length", String.valueOf(full.length));
+            Range.copy(inputStream, outputStream, length, full.start, full.length);
+        } else if (ranges.size() == 1) {
+            // Return single part of file.
+            Range r = ranges.get(0);
+            LOGGER.info("Return 1 part of file : from (" + r.start + ") to (" + r.end + ")");
```

Comment:
“Return 1 part” reads like “Return part nr. one”. Maybe change to “returning part of file”